src/bin/cratesfyi.rs

```diff
@@ -550,39 +550,31 @@ impl BinContext {
     }
 }
 
-impl Context for BinContext {
-    fn build_queue(&self) -> Result<Arc<BuildQueue>, Error> {
-        Ok(self
-            .build_queue
-            .get_or_try_init::<_, Error>(|| {
-                Ok(Arc::new(BuildQueue::new(
-                    self.pool()?,
-                    self.metrics()?,
-                    &*self.config()?,
-                )))
-            })?
-            .clone())
+macro_rules! lazy {
+    (fn $name:ident($self:ident) -> $type:ty = $init:expr) => {
+        fn $name(&$self) -> Result<Arc<$type>, Error> {
+            Ok($self
+                .$name
+                .get_or_try_init::<_, Error>(|| Ok(Arc::new($init)))?
+                .clone())
+        }
     }
+}
 
-    fn storage(&self) -> Result<Arc<Storage>, Error> {
-        Ok(self
-            .storage
-            .get_or_try_init::<_, Error>(|| {
-                Ok(Arc::new(Storage::new(
-                    self.pool()?,
-                    self.metrics()?,
-                    &*self.config()?,
-                )?))
-            })?
-            .clone())
-    }
+impl Context for BinContext {
+    lazy! { fn build_queue(self) -> BuildQueue = BuildQueue::new(
+        self.pool()?,
+        self.metrics()?,
+        &*self.config()?,
+    ) }
 
-    fn config(&self) -> Result<Arc<Config>, Error> {
-        Ok(self
-            .config
-            .get_or_try_init::<_, Error>(|| Ok(Arc::new(Config::from_env()?)))?
-            .clone())
-    }
+    lazy! { fn storage(self) -> Storage = Storage::new(
+            self.pool()?,
+            self.metrics()?,
+            &*self.config()?,
+    )? }
+
+    lazy! { fn config(self) -> Config = Config::from_env()? }
 
     fn pool(&self) -> Result<Pool, Error> {
```

Comment:
This one can't be part of the macro because it doesn't return `Arc`. I thought it was simpler to just leave it be than to try and make the macro more flexible.